xbmc/utils/ExtensionManager.cpp

```diff
@@ -0,0 +1,142 @@
+ *  version 2.1 of the License, or (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *  Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public
+ *  License along with this library; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ */
+
+#include "ExtensionManager.h"
+
+#include <string>
+#include <vector>
+
+#include "addons/AddonManager.h"
+#include "addons/binary-addons/BinaryAddonBase.h"
+#include "ServiceBroker.h"
+
+using namespace ADDON;
+
+const std::vector<TYPE> ADDON_TYPES = {
+  ADDON_VFS,
+  ADDON_IMAGEDECODER,
+  ADDON_AUDIODECODER
+};
+
+CExtensionManager::CExtensionManager()
+{
+  m_advancedSettings = g_advancedSettingsRef;
+
+  InitAddonExtensions();
+
+  CServiceBroker::GetAddonMgr().Events().Subscribe(this, &CExtensionManager::OnAddonEvent);
+}
+
+CExtensionManager::~CExtensionManager()
+{
+  CServiceBroker::GetAddonMgr().Events().Unsubscribe(this);
+
+  m_advancedSettings.reset();
+  m_addonExtensions.clear();
+}
+
+std::string CExtensionManager::GetDiscStubExtensions() const
+{
+  return m_advancedSettings->m_discStubExtensions;
+}
+
+std::string CExtensionManager::GetMusicExtensions() const
+{
+  std::string extensions(m_advancedSettings->m_musicExtensions);
+  extensions += '|' + GetAddonExtensions(ADDON_VFS);
+  extensions += '|' + GetAddonExtensions(ADDON_AUDIODECODER);
+
+  return extensions;
+}
+
+std::string CExtensionManager::GetPictureExtensions() const
+{
+  std::string extensions(m_advancedSettings->m_pictureExtensions);
+  extensions += '|' + GetAddonExtensions(ADDON_VFS);
+  extensions += '|' + GetAddonExtensions(ADDON_IMAGEDECODER);
+
+  return extensions;
+}
+
+std::string CExtensionManager::GetSubtitleExtensions() const
+{
+  std::string extensions(m_advancedSettings->m_subtitlesExtensions);
+  extensions += '|' + GetAddonExtensions(ADDON_VFS);
+
+  return extensions;
+}
+
+std::string CExtensionManager::GetVideoExtensions() const
+{
+  std::string extensions(m_advancedSettings->m_videoExtensions);
+  extensions += '|' + GetAddonExtensions(ADDON_VFS);
+
+  return extensions;
+}
+
+std::string CExtensionManager::GetAddonExtensions(const TYPE &type) const
+{
+  auto it = m_addonExtensions.find(type);
+  if (it != m_addonExtensions.end())
+    return it->second;
+
+  return "";
+}
+
+void CExtensionManager::InitAddonExtensions()
+{
+  for (auto const type : ADDON_TYPES)
+  {
+    SetAddonExtensions(type);
+  }
+}
+
+void CExtensionManager::SetAddonExtensions(const TYPE& type)
+{
+  std::vector<std::string> extensions;
+  BinaryAddonBaseList addonInfos;
+  CServiceBroker::GetBinaryAddonManager().GetAddonInfos(addonInfos, true, type);
+  for (auto addonInfo : addonInfos)
```

Comment:
You can use a `for (const auto& addonInfo : addonInfos)` to prevent a class copy

Reply:
sure

Reply:
done